File: word-break/Solution139.java
// https://leetcode.com/problems/word-break/

import java.util.*;

class Solution139 {

    // Memoization map to store computed results for specific starting indices
    // Key: starting index in the string
    // Value: whether the substring from this index can be broken into dict words
    private Map<Integer, Boolean> memo;

    // T: O(n*m*t), S: O(n)
    public boolean wordBreak(String s, List<String> wordDict) {
        // Initialize memoization map
        // Add base case: empty string at the end of the string is always breakable
        memo = new HashMap<>();
        memo.put(s.length(), true);

        // Start depth-first search from the beginning of the string
        return dfs(s, wordDict, 0);
    }

    private boolean dfs(String s, List<String> wordDict, int i) {
        // If we've already computed result for this starting index, return it
        if (memo.containsKey(i)) {
            return memo.get(i);
        }

        // Try to match words from the dictionary at the current index
        for (String w : wordDict) {
            // Check if the current word fits within the remaining string
            if (i + w.length() <= s.length() &&
            // Check if the substring matches the current word
                    s.substring(i, i + w.length()).equals(w)) {
                // Recursively check if the rest of the string can be broken
                if (dfs(s, wordDict, i + w.length())) {
                    // If a valid word break is found, memoize and return true
                    memo.put(i, true);
                    return true;
                }
            }
        }

        // If no valid word break is found from this index, memoize and return false
        memo.put(i, false);
        return false;
    }

    public static void main(String[] args) {
        var sol = new Solution139();

        String s1 = "leetcode";
        List<String> wordDict1 = Arrays.asList("leet", "code");
        System.out.println(sol.wordBreak(s1, wordDict1)); // true

        String s2 = "applepenapple";
        List<String> wordDict2 = Arrays.asList("apple", "pen");
        System.out.println(sol.wordBreak(s2, wordDict2)); // true

        String s3 = "catsandog";
        List<String> wordDict3 = Arrays.asList("cats", "dog", "sand", "and", "cat");
        System.out.println(sol.wordBreak(s3, wordDict3)); // false

        String s4 = "";
        List<String> wordDict4 = Arrays.asList("a", "b");
        System.out.println(sol.wordBreak(s4, wordDict4)); // true

        String s5 = "hello";
        List<String> wordDict5 = Arrays.asList("hello");
        System.out.println(sol.wordBreak(s5, wordDict5)); // true

        String s6 = "aaaaaaa";
        List<String> wordDict6 = Arrays.asList("aaaa", "aaa");
        System.out.println(sol.wordBreak(s6, wordDict6)); // true
    }

}

File: word-break/solution139.go
// https://leetcode.com/problems/word-break/

package main

import (
	"fmt"
)

// T: O(n*m*t), S: O(n)
func wordBreak(s string, wordDict []string) bool {
	memo := make(map[int]bool)
	memo[len(s)] = true
	return dfs(s, wordDict, 0, memo)
}

func dfs(s string, wordDict []string, i int, memo map[int]bool) bool {
	if val, ok := memo[i]; ok {
		return val
	}

	for _, w := range wordDict {
		if i+len(w) <= len(s) && s[i:i+len(w)] == w {
			if dfs(s, wordDict, i+len(w), memo) {
				memo[i] = true
				return true
			}
		}
	}

	memo[i] = false
	return false
}

func main() {
	s1 := "leetcode"
	wordDict1 := []string{"leet", "code"}
	fmt.Println(wordBreak(s1, wordDict1)) // true

	s2 := "applepenapple"
	wordDict2 := []string{"apple", "pen"}
	fmt.Println(wordBreak(s2, wordDict2)) // true

	s3 := "catsandog"
	wordDict3 := []string{"cats", "dog", "sand", "and", "cat"}
	fmt.Println(wordBreak(s3, wordDict3)) // false

	s4 := ""
	wordDict4 := []string{"a", "b"}
	fmt.Println(wordBreak(s4, wordDict4)) // true

	s5 := "hello"
	wordDict5 := []string{"hello"}
	fmt.Println(wordBreak(s5, wordDict5)) // true

	s6 := "aaaaaaa"
	wordDict6 := []string{"aaaa", "aaa"}
	fmt.Println(wordBreak(s6, wordDict6)) // true
}
